### DISCUSSION TITLE
Forecast with GRE over IPsec

### DISCUSSION THREAD

**Discussion by michaelinoue** [strongswan/strongswan#2365] [Q&A] [answered]
I am trying to use forecast with GRE over IPsec in transport mode. I did not foresee any reasons why this would be problematic (the [example](https://www.strongswan.org/testing/testresults/ikev2/host2host-transport-connmark/) provided does not use GRE but it does use transport mode), but I have yet to be able to form a route over the GRE tunnels I'm creating (doing a deep dive with pwru in fact demonstrates that my packets are getting freed at egress in ip_tunnel_xmit). Without forecast enabled this is also an issue; it seems like it's the `%unique `packet marks itself that is the problem (all things being equal, when I don't enable the `%unique` packet marks, I can form routes, when it is enabled, I cannot).

Forgive my ignorance, but could there be any problems with this GRE over IPsec setup in general with respect to the packet marking that is required for forecast?

**Comment by tobiasbrunner** [comment]
> Without forecast enabled this is also an issue; it seems like it's the `%unique `packet marks itself that is the problem (all things being equal, when I don't enable the `%unique` packet marks, I can form routes, when it is enabled, I cannot).

That will definitely be an issue as without the plugin, there is nothing marking the packets to match the unique marks on IPsec policies/SAs. The plugin installs firewall rules to do so for packets that match the traffic selectors (note that it only works with IPv4 and the rules are IP/subnet based, so no protocol/port selectors are currently applied to the rules).

> Forgive my ignorance, but could there be any problems with this GRE over IPsec setup in general with respect to the packet marking that is required for forecast?

Definitely possible. Could you please provide some more details on your config, logs and firewall dumps (see #196). Also, what exactly do you mean with "form a route"?

**Reply by michaelinoue** [reply]
> Definitely possible. Could you please provide some more details on your config, logs and firewall dumps (see #196). Also, what exactly do you mean with "form a route"?

I am running a pretty proprietary setup, so there's a lot I cannot share (apologies for missing the before-you-post, I will try to shift my question to be more general). However, I will try to reproduce the issue in a more generic setup that I can share if possible. For what it's worth, `pwru` indicates that my packet hasn't even been encapsulated with a GRE header when the packet gets dropped (as mentioned, it is dropped in `ip_tunnel_xmit` in the linux kernel). By "form a route" I just mean that my routing daemon isn't able to send hello messages for route discovery as its packets cannot leave my GRE interface.


What I'm really wondering is if GRE tunnels would be intrinsically incompatible with the forecast plugin due to the marking behavior that's required post-encapsulation. I would assume the PREROUTING rules for marking the packets for the src/dst of the GRE interfaces would take care of this, but am unsure whether the marks that would be required would be applied "too late", i.e. not marking by the time the GRE tunnel expects it for routing and leading to a packet that doesn't match a security policy and subsequently getting dropped. Seems like at least for my setup, that is the case that there is no route for the GRE interface:

```# ip -s tunnel show
gre0: gre/ip remote any local any ttl inherit nopmtudisc
RX: Packets    Bytes        Errors CsumErrs OutOfSeq Mcasts
    0          0            0      0        0        0
TX: Packets    Bytes        Errors DeadLoop NoRoute  NoBufs
    0          0            0      0        0        0
my-intf: gre/ip remote 192.168.10.5 local 192.168.10.1 ttl 255 nopmtudisc key 2658837050
RX: Packets    Bytes        Errors CsumErrs OutOfSeq Mcasts
    0          0            0      0        0        0
TX: Packets    Bytes        Errors DeadLoop NoRoute  NoBufs
    0          0            4432   0        4432     0
```

If they are compatible, i.e. it is acceptable to rely on forecast marking GRE packets and that should suffice for GRE tunneling in transport mode, then I will dig deeper into what could be going wrong with my setup.

**Reply by tobiasbrunner** [reply]
> I would assume the PREROUTING rules for marking the packets for the src/dst of the GRE interfaces 

Why is marking required for GRE?

> Seems like at least for my setup, that is the case that there is no route for the GRE interface

What do you mean?

I've not thought about this too much before, but now I don't really see how this combination even makes sense. The policy for GRE over IPsec is host-to-host transport mode with GRE protocol selector, while the [**forecast** plugin](https://docs.strongswan.org/docs/5.9/plugins/forecast.html) is intended to forward broad- and multicast packets over a tunnel mode SA with matching selectors, i.e. that will never work with such a specific  policy. You basically have to get multicast packets into the GRE tunnel, which the _forecast_ plugin can't help you with. I also didn't notice before that you actually referenced an example that uses the [**connmark** plugin](https://docs.strongswan.org/docs/5.9/plugins/connmark.html), which is something completely different. Are you maybe confusing the two plugins? If so, what exactly would be the use case for the _connmark_ plugin here?

**Reply by michaelinoue** [reply]
>Why is marking required for GRE?

It's not, but it is required for the forecast plugin. I'm afraid that what might be happening is that marks are not applied to the newly encapsulated GRE packet and are hence getting dropped. I was referring to the PREROUTING rules earlier; my understanding was that before transmitting the packet out of the GRE interface, that the marks would be applied by forecast. Indeed, looking at the PREROUTING rules it does add a mark for packets leaving the GRE interface. When I was saying "in time" I just mean that I'm unsure if the kernel is dropping these packets before the PREROUTING rules have even been applied. 

>Seems like at least for my setup, that is the case that there is no route for the GRE interface
>>What do you mean?

What I mean is that GRE packets are getting dropped at transmit. I suspect this is because packet marking has led to packets being dropped.

>You basically have to get multicast packets into the GRE tunnel, which the forecast plugin can't help you with.

My current setup is a GRE interface, with security policies using the point-to-point GRE interface addresses. So any traffic using the GRE interface would get encapsulated with that address and hence match the security policies as they explicitly use the GRE addresses. Previously, without any marks added, this worked. The routing discovery protocol I was using sent out a "hello" message over the GRE interface, would get encapsulated, match the security policy and get encrypted, arrive at the other end of the GRE interface, get decrypted and decapsulated, and would get a "hello" response in the same fashion, providing a route to my other host via the GRE interface. Fundamentally my problem is just getting traffic over the GRE interface now, not even getting multicast traffic forwarded.

>The policy for GRE over IPsec is host-to-host transport mode with GRE protocol selector, while the [forecast plugin](https://docs.strongswan.org/docs/5.9/plugins/forecast.html) is intended to forward broad- and multicast packets over a tunnel mode SA with matching selectors, i.e. that will never work with such a specific policy.

So this is just a fundamental problem with using GRE transport mode? i.e. even if I am able to establish a route over the GRE interface, forecast won't be able to forward multicast packets as it strictly depends on using IPsec tunnel mode? 
>I also didn't notice before that you actually referenced an example that uses the [connmark plugin](https://docs.strongswan.org/docs/5.9/plugins/connmark.html), which is something completely different. Are you maybe confusing the two plugins? If so, what exactly would be the use case for the connmark plugin here?

I'm not using connmark, that was just an example I found linked to the forecast documentation.

**Reply by tobiasbrunner** [reply, marked as answer]
> I was referring to the PREROUTING rules earlier; my understanding was that before transmitting the packet out of the GRE interface, that the marks would be applied by forecast.

Why PREROUTING?  I'd assume for the locally generated GRE packets it would be the rules in OUTPUT if anything (only forwarded plaintext packets would go through PREROUTING, but before hitting the GRE interface).

> Fundamentally my problem is just getting traffic over the GRE interface now, not even getting multicast traffic forwarded.

Then just disable the unique marks. Again, the combination of GRE tunneling and the _forecast_ plugin really doesn't make much sense, so you don't need any unique marks.

> i.e. even if I am able to establish a route over the GRE interface, forecast won't be able to forward multicast packets as it strictly depends on using IPsec tunnel mode?

Yeah, those things have absolutely nothing to do with each other. GRE and IPsec are completely independent. And really the only reason to use the  _forecast_ plugin is to copy broad-/multicast packets and re-send them with different marks applied so they get tunneled into the tunnel mode SAs with different clients, using IPsec policy matching with those unique marks. This method won't get such packets into GRE interfaces.
